feat(providers): Add OpenID Connect dynamic provider

smkent opened the pull request:
# Submitting Pull Requests

## General

 - [x] Make sure you use [semantic commit messages](https://seesparkbox.com/foundry/semantic_commit_messages).
       Examples: `"fix(google): Fixed foobar bug"`, `"feat(accounts): Added foobar feature"`.
 - [x] All Python code must formatted using Black, and clean from pep8 and isort issues.
 - [x] JavaScript code should adhere to [StandardJS](https://standardjs.com).
 - [x] If your changes are significant, please update `ChangeLog.rst`.
 - [x] If your change is substantial, feel free to add yourself to `AUTHORS`.

 ## Provider Specifics

 In case you add a new provider:

- [x] Make sure unit tests are available.
- [x] Add an entry of your provider in `test_settings.py::INSTALLED_APPS` and `docs/installation.rst::INSTALLED_APPS`.
- [x] Add documentation to `docs/providers.rst`.
- [x] Add an entry to the list of supported providers over at `docs/overview.rst`.

smkent commented:
To see a working concept of this change in action, [see my PR to add OpenID Connect support to paperless-ngx](https://github.com/paperless-ngx/paperless-ngx/pull/1746).

bufke commented:
This looks amazing. Not my call, but I think the need to add a list in settings.py is cumbersome and would be better solved as a database value. This was my suggestion in #3072. Since you're putting the client id and secret in settings, it makes the "social application" model entirely redundant. I see why you did it this way as it's not possible to configure multiple of the same provider.

IMO if it was moved to a modified version of the social application model, it would be easier to configure and allow a developer to enable end user uploaded oidc themselves which would be good for a SaaS app.

smkent commented:
@bufke I'm glad you like my change!

django-allauth actually supports both database-based configuration in the social applications DB model or client IDs/keys in provider `APP` values for existing providers. This is mentioned at the [bottom of the long example block in the Installation -> Django section of the docs](https://django-allauth.readthedocs.io/en/latest/installation.html#django).

Similarly, this change also supports configuration via the database as an option instead of the `APP` config var method. For example, config like this:

```python
SOCIALACCOUNT_PROVIDERS = {
    "openid-connect": {
        "SERVERS": [
            dict(
                id="example_oidc",
                name="Example OIDC",
                server_url="https://oidc.example.com",
            ),
        ],
    },
}
```

adds a new provider to social application providers list:

![scr](https://user-images.githubusercontent.com/2831985/197917577-995dbc82-0d0a-4ac7-b20b-3f98a8a59675.png)

bufke commented:
Interesting! That's from _provider_factory, yes? And not typical django-allauth behavior. That mitigates 3072 - though IMO it's still not preferable for SaaS. Secret and client ID are editable in the DB, but not other configuration values. Such as a provider tenant name or in this case, server url.

I'll give your code a test on some services this week or next. It solves a few problems for an open source project I'm affiliated with - [KoboToolbox](kobotoolbox.org/). We're migrating to allauth and need to support potentially many thousands of organizations. That's why reading them in settings.py for each one isn't as nice as doing it Django Admin.

smkent commented:
> That's from _provider_factory, yes?

Right, `_provider_factory` dynamically creates a new provider instance for each configured OpenID Connect server.

> And not typical django-allauth behavior.

Logically, OpenID Connect is a protocol (similar to, and in fact built on top of OAuth2). django-allauth exposes implementations of protocols as providers, such as Google (which uses OAuth2) or Keycloak (which uses OpenID Connect in the existing codebase). 

I wanted to keep this change consistent with the existing design patterns. To support arbitrary OpenID Connect providers and keep them as logical top-level objects like the other providers, I'm using a factory function to create independent classes for each configured OpenID Connect provider.

> That mitigates 3072 - though IMO it's still not preferable for SaaS. Secret and client ID are editable in the DB, but not other configuration values. Such as a provider tenant name or in this case, server url.

Your use case in #3072 is definitely reasonable, and django-allauth would likely benefit from more configuration flexibility. IMO a change on that front should be separate from a change like this, which simply adds new provider(s) matching the existing design pattern.

> I'll give your code a test on some services this week or next. It solves a few problems for an open source project I'm affiliated with - [KoboToolbox](kobotoolbox.org/). We're migrating to allauth and need to support potentially many thousands of organizations. That's why reading them in settings.py for each one isn't as nice as doing it Django Admin.

Thanks for giving my change a try and I hope it works well!

bufke commented:
I tested with Google Apps and Azure and it worked great! I can add multiple Azure tenants without resorting to any hacks like creating duplicate provider classes! Azure's server url has the tenant ID in it already.

The instructions made sense to me and explained the unique oidc/ID callback url. I agree you shouldn't blow up the scope of this PR.

bufke commented:
Tested with OneLogin but couldn't get it working. I'll try to debug. They seem to support OIDC. Their server URL is https://<tenant name>.onelogin.com/oidc/2

smkent commented on allauth/socialaccount/providers/openid_connect/views.py:
```diff
@@ -0,0 +1,40 @@
+        return self.openid_config["userinfo_endpoint"]
+
+    def complete_login(self, request, app, token, response):
+        response = requests.post(
```
Great catch, this should definitely be a `GET` request. [The existing provider code for Keycloak makes a `POST` request for user info](https://github.com/pennersr/django-allauth/blob/69e571f89e5b98cb2eb308b8b788532d6746f0b6/allauth/socialaccount/providers/keycloak/views.py#L33), which is how this ended up in my change. I've updated this PR to use a `GET` request instead.

smkent replied:
The `master` branch now performs a `GET` request on the userinfo endpoint in the Keycloak provider as of 39f8cafbf58c98f08193eb353e117405330abb14. Using `GET` here is no longer new functionality in this PR.

smkent commented:
Hi @pennersr -- just a friendly ping! It would be great to get this merged. Generalized OpenID Connect support would make django-allauth useful for a number of applications such as paperless-ngx and Tandoor Recipes. I'm more than happy to consider any of your feedback. Thanks!

smkent commented:
I've resolved the merge conflict on `ChangeLog.rst`.

kurokay commented:
Nice, i'm waiting this feature to be merged. Will be awesome !

muhmuhhum commented:
I have a small little problem when i want to set the the OAUTH_PKCE_ENABLED setting for a openId_connect social account.
`
  "openid-connect'" {
              "SERVERS": [
                  {
                      "id": "testAuth",
                      "name": "Test Auth",
                      "server_url": SOCIAL_OIDC_URL,
                      "APP": {
                          'client_id': SOCIAL_OIDC_CLIENT_ID,
                          'secret': SOCIAL_OIDC_SECRET
                      }
                  }
              ]
          }
`

For this i have to make a workaround
`
'openid-connect_testAuth':{
            'OAUTH_PKCE_ENABLED': True
        }
`

smkent commented:
> I have a small little problem when i want to set the the OAUTH_PKCE_ENABLED setting for a openId_connect social account. `"openid-connect'" { "SERVERS": [ { "id": "testAuth", "name": "Test Auth", "server_url": SOCIAL_OIDC_URL, "APP": { 'client_id': SOCIAL_OIDC_CLIENT_ID, 'secret': SOCIAL_OIDC_SECRET } } ] }`
> 
> For this i have to make a workaround `'openid-connect_testAuth':{ 'OAUTH_PKCE_ENABLED': True }`

@muhmuhhum Thanks for letting me know! I've updated how the OpenID Connect per-server settings are accessed internally and added a corresponding unit test. It should now be possible to set `OAUTH_PKCE_ENABLED` like this:

```python
"openid-connect": {
    "SERVERS": [
        {
            "id": "testAuth",
            "name": "Test Auth",
            "server_url": SOCIAL_OIDC_URL,
            "OAUTH_PKCE_ENABLED": True,
            "APP": {
                "client_id": SOCIAL_OIDC_CLIENT_ID,
                "secret": SOCIAL_OIDC_SECRET,
            },
        }
    ]
}
```

andrey-toptal commented:
any review updates on this PR?

mikapfl commented:
Hi,

When testing this PR, I encountered two difficulties. I configured an oidc provider (a self-hosted Ory Hydra), using a config like (anonymized):
```
SOCIALACCOUNT_PROVIDERS = {
    'openid-connect': {
        'SERVERS': [
            dict(
                id='oidc.servers.tld',
                name="Custom Account System",
                server_url="https://accounts.example.org",
                APP=dict(
                    client_id="example",
                    secret="supersecure",
                ),
            ),
        ]
    }
}
```

Difficulties:
1. Apparently, the implementation uses `client_secret_post` as a token endpoint authentication method. In the default configuration, my OIDC Provider did not allow this. Maybe document this somewhere?
2. The generated provider ID was too long and wouldn't fit into the database, and I got a `django.db.utils.DataError: value too long for type character varying(30)` exception. Apparently, the provider factory adds at least 15 characters of provider ID because my id was 16 characters long. After switching to an ID with only 3 characters, things started working.

Debugging these problems was rather challenging because even with `DEBUG = True` in the django settings, no debugging information is displayed by the [authentication_error template](https://github.com/pennersr/django-allauth/blob/master/allauth/templates/socialaccount/authentication_error.html). I had to patch the template to find about the token endpoint authentication method mismatch.

After debugging and fixing these two issues, the provider works like a charm and pulls information. One final thing I noticed is that the `email_verified` claim in the oidc spec is not used. It could be used to mark the email as verified already and not send out the verification email.

Thanks for your work!

Cheers

Mika

smkent commented:
@mikapfl Thank you for your detailed feedback!

I've added some changes:

* I noted the provider ID 15-character limit in the documentation's configuration example. We could gain a few more characters if the `openid-connect_` internal ID prefix is changed to something shorter. I'm open to doing that if folks feel strongly.
* The OAuth2 provider base (which the OpenID Connect provider inherits from) supports token retrieval via both HTTP Basic authentication and HTTP POST data. I added support for both options in the OpenID Connect provider. Each server can now have an optional `token_auth_method` value (`client_secret_basic` or `client_secret_post`) to use a specific method. Otherwise, a method is selected from the server's `token_endpoint_auth_methods_supported` list in the `.well-known/openid-configuration` response. (This new behavior also applies to the Keycloak provider since it inherits from OpenID Connect.)
* I added support for the `email_verified` claim in the userinfo response by implementing `extract_email_addresses` on the provider.

Please let me know if you encounter any more issues with the OpenID Connect provider.

smkent commented:
Hi @pennersr -- just another friendly ping! A number of folks are finding this feature useful and would benefit from this PR being merged and included in a new release. Please let me know if you have any review feedback and/or what else I can do to help move this feature forward. Thank you!

smkent commented:
With @pennersr's announcement today in #3201 regarding their limited time for django-allauth, I'm not sure what the timeline might look like for getting this feature merged and released. I've started a discussion about making a continuity plan for django-allauth in https://github.com/pennersr/django-allauth/issues/3202 and would love to hear people's thoughts.

pennersr commented on allauth/socialaccount/providers/openid_connect/provider.py:
```diff
@@ -0,0 +1,91 @@
+
+
+class OpenIDConnectProvider(OAuth2Provider):
+    id = "openid-connect"
```
The convention is that the `id` matches the package name, `openid_connect` in this case. Though, perhaps it would be best to rename the package (and the `id`) to `oidc`?

smkent replied:
I changed the provider ID to `openid_connect` as suggested. I'm happy to change it to `oidc` if folks feel strongly. If we were to rename to `oidc`, the `openid_connect` provider directory should also be renamed to `oidc`.

pennersr commented on allauth/socialaccount/providers/openid_connect/urls.py:
```diff
@@ -0,0 +1,45 @@
+        ),
+    ]
+
+    return [path(provider.get_slug() + "/", include(urlpatterns))]
```
I am not sure we should nest OIDC based providers below a  `/accounts/oidc/` path. The fact that such providers are using this subprovider mechanism is just an implementation detail that better not shine through on the outside.  Not having the "/oidc/" in between would make it easier to move to (or perhaps even away from) an OIDC based subprovider.

smkent replied:
I have removed the `/oidc` component from the OpenID Connect URL paths.

smkent replied:
Note that this is a change to the OpenID Connect provider's `get_slug` method via https://github.com/pennersr/django-allauth/pull/3165/commits/cc77a03aecf1ad68caac10753bba40e2ed075318.

pennersr commented on allauth/socialaccount/providers/openid_connect/views.py:
```diff
@@ -0,0 +1,49 @@
+        )
+        response.raise_for_status()
+        extra_data = response.json()
+        extra_data["id"] = extra_data["sub"]
```
This looks like we're deviating from the standard -- is it really needed to do this by default, or just a mechanism to cope with legacy situations?

smkent replied:
I have removed this. The OpenID Connect provider's `extract_uid` method now simply consumes `extra_data["sub"]`. 

[As described in the spec, `sub` in the userinfo endpoint response contains the user ID.](https://openid.net/specs/openid-connect-core-1_0.html#UserInfoResponse)

pennersr commented on docs/providers.rst:
```diff
@@ -1485,6 +1485,54 @@ The OpenID provider can be forced to operate in stateless mode as follows::
+    SOCIALACCOUNT_PROVIDERS = {
+        'openid-connect': {
+            'SERVERS': [
+                dict(
```
It would be more readable to stick to one way of defining structures (so `{` instead of `dict(`).

smkent replied:
I have replaced `dict()` constructor usage with `{}` literals.

smkent commented:
Hi @pennersr, thank you very much for your feedback! I have added updates based on your comments. This PR is now ready for another look.

pennersr commented:
When you add an OpenID connect provider with id `foo` the resulting `SocialAccount.provider` field will be filled with `openid_connect_foo`. This is problematic because that field has a max length of 30, 15 of those get consumed by `openid_connect_`.  I would prefer a shorter prefix, and also use a different separator, e.g. `oidc:foo`.  

Alternatively, you can also argue that we should not expose internal implementation details of provider `foo` at the database level, and just store `foo`. 

Currently this PR only supports configuration via the settings. We could/should extend this to also support `SocialApp` based configuration.  We don't need to do that as part of this PR, but if we were to pick that up, I think it would look something like this:

- `SocialApp` needs a new URL field (for `server_url`).
- We would need to turn `SocialApp.provider` into a free format text field.
- In that field, you would need to enter `oidc:foo`, or perhaps friendlier, add another new field `type` that would be filled with `oidc`, and set `provider` to just `foo`.

All things considered, I think having just `foo` makes most sense. What do you think?

smkent commented:
> When you add an OpenID connect provider with id `foo` the resulting `SocialAccount.provider` field will be filled with `openid_connect_foo`. This is problematic because that field has a max length of 30, 15 of those get consumed by `openid_connect_`. I would prefer a shorter prefix, and also use a different separator, e.g. `oidc:foo`.
> 
> Alternatively, you can also argue that we should not expose internal implementation details of provider `foo` at the database level, and just store `foo`.
> 
> [...]
>
> All things considered, I think having just `foo` makes most sense. What do you think?

@pennersr Removing the `openid_connect_` prefix works for me. I removed it and updated the PR.

> Currently this PR only supports configuration via the settings. We could/should extend this to also support `SocialApp` based configuration. We don't need to do that as part of this PR, but if we were to pick that up, I think it would look something like this:
> 
>     * `SocialApp` needs a new URL field (for `server_url`).
> 
>     * We would need to turn `SocialApp.provider` into a free format text field.
> 
>     * In that field, you would need to enter `oidc:foo`, or perhaps friendlier, add another new field `type` that would be filled with `oidc`, and set `provider` to just `foo`.

I agree that this is a great conversation to have separately.

pennersr commented on docs/providers.rst:
```diff
@@ -1486,6 +1486,54 @@ The OpenID provider can be forced to operate in stateless mode as follows::
+        "openid_connect": {
+            "SERVERS": [
+                {
+                    "id": "my-server",  # 15 characters or less
```
**Issue**: By dropping the prefix, this 15 characters remark and the one below can be bumped to 30.

smkent replied:
I updated both comments to say 30 characters or less.

pennersr commented:
Merged, thanks!

smkent commented:
@pennersr Thank you so much! I'm eager to consume this functionality in some other projects. Would you be able to release a new version of django-allauth? Thanks again!

em- commented:
Thank you both for implementing and shipping this! I am pretty sure we will be playing with this soon. :D
